### PR TITLE
[SPARK-13225] [SQL] Support Intersect All/Distinct [WIP]

### PR DESCRIPTION
**Pull request by gatorsmile**
In the SQL2003 Syntax, INTERSECT supports both ALL and DISTINCT

```
<SELECT statement1>
INTERSECT [ALL | DISTINCT]
<SELECT statement2>
```

This PR is to support both. 
- [x] Enable both optional key words `ALL` and `DISTINCT` when parsing `INTERSECT` in SQL Parser;
- [x] Add a new option `INTERSECT ALL`. It avoids adding extra `DISTINCT` above `Left-semi JOIN` after conversion from `INTERSECT` to `Left-semi JOIN`. 
- [ ] Add the corresponding APIs of Dataframe and Dataset.

### CODE REVIEW

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g [hunk 1]
```diff
@@ -2222,7 +2223,9 @@ setOperator
     : KW_UNION KW_ALL -> ^(TOK_UNIONALL)
     | KW_UNION KW_DISTINCT? -> ^(TOK_UNIONDISTINCT)
     | KW_EXCEPT -> ^(TOK_EXCEPT)
-    | KW_INTERSECT -> ^(TOK_INTERSECT)
+    | KW_INTERSECT (all=KW_ALL | distinct=KW_DISTINCT)?
+    -> {$all == null}? ^(TOK_INTERSECTDISTINCT)
+    ->                 ^(TOK_INTERSECTALL)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g]: So distinct is optional?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -1059,19 +1059,24 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
...
- * 2. This rule has to be done after de-duplicating the attributes; otherwise, the generated
+ * Note: This rule has to be done after de-duplicating the attributes; otherwise, the generated
  *    join conditions will be incorrect.
  */
 object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case Intersect(left, right) =>
+    case Intersect(left, right, distinct) =>
       assert(left.output.size == right.output.size)
       val joinCond = left.output.zip(right.output).map { case (l, r) => EqualNullSafe(l, r) }
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I think you actually use ```EqualNullSafe.tupled``` here and map over that.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 3]
```diff
@@ -594,8 +594,9 @@ class Dataset[T] private[sql](
    * and thus is not affected by a custom `equals` function defined on `T`.
    * @since 1.6.0
    */
-  def intersect(other: Dataset[T]): Dataset[T] = withPlan[T](other)(Intersect)
-
+  def intersect(other: Dataset[T]): Dataset[T] = withPlan[T](other){ (left, right) =>
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Nit: no-space between ){

**Comment by hvanhovell** [comment]
retest this please

File: sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g (hunk 1 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/antlr3/org/apache/spark/sql/catalyst/parser/SparkSqlParser.g]: Yeah, the default behavior of `INTERSECT` is `INTERSECT DISTINCT`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Sure, will change it. Thanks!

**Comment by gatorsmile** [comment]
@marmbrus @rxin I want to get your opinions first before adding the corresponding Dataframe and Dataset APIs.

Now, `unionall` is changed to `union`, whose default behavior is `UNION ALL`. The default behavior of `intersect` is `INTERSECT DISTINCT`. I think we are facing three options:
  - Add a new API named `intersectall`
  - Add an optional boolean parameter `distinct` into the existing `intersect` API, like the APIs `sample` and `repartition`
  - Do not add any API.
 
Thank you!

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -1059,19 +1059,24 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
...
 object ReplaceIntersectWithSemiJoin extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
-    case Intersect(left, right) =>
+    case Intersect(left, right, distinct) =>
       assert(left.output.size == right.output.size)
-      val joinCond = left.output.zip(right.output).map { case (l, r) => EqualNullSafe(l, r) }
-      Distinct(Join(left, right, LeftSemi, joinCond.reduceLeftOption(And)))
+      val joinCond = left.output.zip(right.output).map(EqualNullSafe.tupled)
+      if (distinct) {
+        Distinct(Join(left, right, LeftSemi, joinCond.reduceLeftOption(And)))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Minor: you can move the `joinCond.reduceLeftOption(And)` into the construction of the `joinCond`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 5]
```diff
@@ -223,16 +222,22 @@ object HiveTypeCoercion {
...
+      case s @ Except(left, right) if s.childrenResolved &&
           left.output.length == right.output.length && !s.resolved =>
         val newChildren: Seq[LogicalPlan] = buildNewChildrenWithWiderTypes(left :: right :: Nil)
         assert(newChildren.length == 2)
-        s.makeCopy(Array(newChildren.head, newChildren.last))
+        Except(newChildren.head, newChildren.last)
+
+      case s @ Intersect(left, right, distinct) if s.childrenResolved &&
+          left.output.length == right.output.length && !s.resolved =>
+        val newChildren: Seq[LogicalPlan] = buildNewChildrenWithWiderTypes(left :: right :: Nil)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Minor: there must be exactly two children in the result. We could just do a pattern match on the sequence and be done:

    val Seq(newLeft, newRight) = buildNewChildrenWithWiderTypes(left :: right :: Nil)
    Intersect(newLeft, newRight, distinct)

The same applies to the `Except` case.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Thanks! Will do the changes. : )

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Yeah, let me change it now.

**Comment by rxin** [comment]
I haven't actually looked at your pull request, but I'm fairly sure the implementation is wrong given the number of lines involved. The actual change is probably much larger to implement intersect all.

Intersect all is actually not just a join. It is multisect intersect, e.g.

[1, 2, 2] intersect [1, 2] == [1, 2]
[1, 2, 2] intersect_all [1, 2] == [1, 2]
[1, 2, 2] intersect_all [1, 2, 2] == [1, 2, 2]

i.e. in order to support intersect all, we'd need to count the number of times each row appears.

same thing with except all.

**Comment by gatorsmile** [comment]
Uh, you are right. : ) Will follow your suggestions. Thank you!

**Comment by rxin** [comment]
In terms of API, I think we should just add intersectAll and exceptAll functions to it. 

For union, we should keep the existing behavior, and if users want to do union distinct, they can just do union().distinct().

**Comment by gatorsmile** [comment]
Thank you! Will do it.

**Comment by rxin** [comment]
And can we close this pr and only open it when you have a new version? Thanks.

**Comment by gatorsmile** [comment]
Sure, let me close it first. Will continue to work on it. Thanks!

**Comment by Tagar** [comment]
another possible way to implement INTERSECT ALL
https://issues.apache.org/jira/browse/SPARK-21274

**Pull request closed without merging**